neps/nep-0330.md

```diff
@@ -31,23 +35,36 @@ There is a lot of information that can be held about a contract. Ultimately, we
 
-The metadata will include three optional fields:
+The metadata will include optional fields:
 
-- `version`: a string that references the specific commit hash or version of the code that is currently deployed on-chain. This can be included regardless of whether or not the contract is open-sourced and can also be used for organizational purposes.
+- `version`: a string that references the specific commit hash or version of the code currently deployed on-chain. If the link is to a GitHub repo URL, this is crucial, as it is needed to make the contract reproducible.
 - `link`: a string that references the link to the open-source code. This can be anything such as Github or a CID to somewhere on IPFS.
 - `standards`: a list of objects (see type definition below) that enumerates the NEPs supported by the contract. If this extension is supported, it is advised to also include NEP-330 version 1.1.0 in the list (`{standard: "nep330", version: "1.1.0"}`).
+- `build_details`: a build details object (see type definition below) that contains all the necessary information about how the contract was built, making it possible for others to reproduce the same WASM of this contract.
 
 ```ts
 type ContractSourceMetadata = {
   version: string|null, // optional, commit hash being used for the currently deployed wasm. If the contract is not open-sourced, this could also be a numbering system for internal organization / tracking such as "1.0.0" and "2.1.0".
   link: string|null, // optional, link to open source code such as a Github repository or a CID to somewhere on IPFS.
   standards: Standard[]|null, // optional, standards and extensions implemented in the currently deployed wasm e.g. [{standard: "nep330", version: "1.1.0"},{standard: "nep141", version: "1.0.0"}].
+  build_details: BuildDetails|null, // optional, details that are required for contract WASM reproducibility.
 }
 
 type Standard {
     standard: string, // standard name e.g. "nep141"
     version: string, // semantic version number of the Standard e.g. "1.0.0"
 }
+
+type BuildDetails {
+	env_image: EnvImage, // reference to a reproducible build environment image, e.g. "{tag: "sourcescan/cargo-near:0.6.0", digest: "sha256:bf488476d9c4e49e36862bbdef2c595f88d34a295fd551cc65dc291553849471"}"
+	contract_path: string|null // relative path to contract folder within the source code e.g. "src/contract". Often, it is the root of the repository, so can be omitted. 
+	build_command: string[] // the exact command that was used to build the contract, with all the flags e.g. ["cargo", "near", "build", "no-abi"].
+}
+
+type EnvImage {
+	tag: string // tag of the image e.g. "sourcescan/cargo-near:0.6.0"
+	digest: string // a sha256 hash unique, immutable identifier for a container image, used to verify that the image under the same tag has not changed e.g. "sha256:bf488476d9c4e49e36862bbdef2c595f88d34a295fd551cc65dc291553849471"
```

Comment:
What's the function of the `digest` field? Is it just a long unique number, or is it supposed to be a hash of something specific?

Reply:
@frol So, the `digest` is a hash of a _Docker_ container image on the hub.docker.com, correct?

The word "docker" should be mentioned at least once in the NEP, because it's a bit confusing.

Reply:
> The word "docker" should be mentioned at least once in the NEP, because it's a bit confusing.

Good point. I will clarify this to prevent confusion. Additionally, it's worth noting that the digest might differ across various registries. Therefore, the structure should also be modified as follows:
```ts
type DockerImage {
  registry: string // Docker image registry e.g. "ghcr.io", "hub.docker.com"
  tag: string // tag of the Docker image e.g. "sourcescan/cargo-near:0.6.0"
  digest: string // a sha256 hash unique, immutable identifier for a Docker image, used to verify that the image under the same tag has not changed e.g. "sha256:bf488476d9c4e49e36862bbdef2c595f88d34a295fd551cc65dc291553849471"
}
```

Reply:
35bbcded50a0419a5a24090bd2b0e27b8bcafed5